### ISSUE TITLE
 grpc::ServerContext::TryCancel crashes with segment fault (no matter what!)

### ISSUE THREAD

**Issue by eladm-ultrawis** [grpc/grpc#32638]
I know this issue has be discussed before. But it seems it was dismissed. 

In my case, my service starts listening to client streaming requests like so:

1. initialize grpc::ServerContext.
2. initialize grpc::ServerAsyncWriter with this context
3. start listening (AsyncService::Request<...>)
4. If I call grpc::ServerContext::TryCancel here (from any thread) I get a segmentation fault
5. If I call grpc::ServerContext::TryCancel after the streaming has started I get a segmentation fault as well.

I believe that this method should either be documented more clearly to explain under what scenario it can be called, or at least log / report the reason for the error. A simple fix that would actually cancel the RPC listening from the server side would be nice as well.

I use grpc v1.5.0 on Windows 10 OS.

**Comment by yashykt** [comment]
Hi @eladm-ultrawis, has the RPC actually been made? Can you provide a quick repro?

**Comment by eladm-ultrawis** [comment]
@yashykt The RPC from the client side was not yet initiated. Like I tried to exaplain, the Server service is only listening to client requests. I could not find a way to "cancel" the service listening without shutting down the whole server. calling TryCancel always crashes.

**Comment by yashykt** [comment]
yeah, you are right. The API does not have a way to cancel a `RequestCall`

**Comment by eladm-ultrawis** [comment]
@yashykt thank you for the update, is there a chance to implement the ability to cancel a RequestCall? I think it makes a lot of sense to be able to cancel such a call.

**Comment by yashykt** [comment]
@eladm-ultrawis Please take a look at the async callback API. https://github.com/grpc/grpc/blob/master/examples/cpp/helloworld/greeter_callback_server.cc

**Comment by eladm-ultrawis** [comment]
@yashykt I don't follow.

**Comment by yashykt** [comment]
We are moving to the new callback API world for gRPC C++, and you won't be dealing with some of the issues that you are facing with the CQ based API. Please try out the new API :)

**Comment by eladm-ultrawis** [comment]
@yashykt  I am using asio-grpc which is based on the completion queues. the callback API is not an option. opens up to many threads without being able to control it.

**Comment by Tails86** [comment]
Using async callback API doesn't exactly resolve the root of the issue here. Correct me if I'm wrong, but there doesn't seem to be any way to cancel a request if the server isn't connected yet, even if the callback API is used. The only thing I have found to work around this is to set a deadline in the context so you can reassess on timeout. This polling method seems messy to me because it's difficult to come up with an arbitrary polling period which will work in all cases.
